zeroex/orderwatch/order_watcher.go

```diff
@@ -1157,42 +1159,46 @@ func (w *Watcher) convertValidationResultsIntoOrderEvents(
 			}
 			orderEvents = append(orderEvents, orderEvent)
 		} else {
-			expiration := time.Unix(order.SignedOrder().ExpirationTimeSeconds.Int64(), 0)
+			// The order is expired if ExpirationTimeSeconds is equal to or less than the timestamp
+			// of the validation block.
+			validationBlockTimeStampSeconds := big.NewInt(validationBlock.Timestamp.Unix())
+			isOrderExpired := order.ExpirationTimeSeconds.Cmp(validationBlockTimeStampSeconds) != 1
 
 			if oldFillableAmount.Cmp(newFillableAmount) == 0 {
 				// If order was previously expired, check if it has become unexpired
-				if order.IsRemoved && oldFillableAmount.Cmp(big.NewInt(0)) != 0 && validationBlockTimestamp.Before(expiration) {
-					w.rewatchOrder(order, nil)
+				if order.IsRemoved && oldFillableAmount.Cmp(big.NewInt(0)) != 0 && !isOrderExpired {
+					w.rewatchOrder(order, nil, validationBlock)
 					orderEvent := &zeroex.OrderEvent{
-						Timestamp:                validationBlockTimestamp,
+						Timestamp:                validationBlock.Timestamp,
 						OrderHash:                order.Hash,
 						SignedOrder:              order.SignedOrder(),
 						FillableTakerAssetAmount: order.FillableTakerAssetAmount,
 						EndState:                 zeroex.ESOrderUnexpired,
 					}
 					orderEvents = append(orderEvents, orderEvent)
 				}
-				// No important state-change happened
+				// No important state-change happened. Still want to update lastValidatedBlock
+				w.updateOrderLastValidatedBlock(order, validationBlock)
 				continue
 			}
 			if oldFillableAmount.Cmp(big.NewInt(0)) == 1 && oldAmountIsMoreThenNewAmount {
 				// If order was previously expired, check if it has become unexpired
-				if order.IsRemoved && oldFillableAmount.Cmp(big.NewInt(0)) != 0 && validationBlockTimestamp.Before(expiration) {
-					w.rewatchOrder(order, newFillableAmount)
+				if order.IsRemoved && oldFillableAmount.Cmp(big.NewInt(0)) != 0 && !isOrderExpired {
```

Comment:
It's outside the scope of this PR, but we can simplify this code by checking for unexpiration, and then checking for amount changes.

Reply:
👍 I know this is pretty messy at the moment. Let's try and address it later when we have more time to spend on tech debt.